### ISSUE TITLE
Split "coupling" into: return/raise or not vs. ignore callee failures or not.

### ISSUE THREAD

**Issue by tmontes** [tmontes/python-wires#35]
## Current state

Currently these concepts are intertwined:
* Coupled calls:
  * Stop on first wired callee exception and raise exception with previous callee results.
* Decoupled calls:
  * Call all wired callees, regardless of exceptions, returning list of `(<exception>, <result>)`.

## Idea

Separating these concepts is a sound idea:

|       | `coupling` = `False` | `coupling` = `True` |
|----|----------|-------------|
| `ignore_failures` = `True` | calls all wired callees, absorbing exceptions + returns `None` | calls all wired callees, absorbing exceptions +  returns `(<exception>, <result>)` list |
| `ignore_failures` = `False` | stops calling wired callees at the 1st exception + returns `None`  | stops calling wired callees at the 1st exception + raises exception wrapping `(<exception>, <result>)` list |

PS: This approach can also support setting `ignore_failures` to a non-negative integer value, defining the number of failures to ignore, before stopping.

## API thoughts

Instance creation API is easily adaptable; it's a matter of using separate initialisation arguments:
```
Wiring(coupling=..., ignore_failures=...)
```

Call-time overriding is not so obvious. Currently call-time overriding is done via:
```
w.coupled.this()
w.decoupled.this()
```

How to go about and combine `coupling` and  `ignore_failures` overriding? What about possible future settings/overrides?

### Ideas

**Context manager?**

```
with w.override(coupling=..., ignore_failures=...) as w:
    w.this()
```

Or even shorter:

```
with w(coupling=..., ignore_failures=...) as w
    w.this()
```

**Parameterized attribute?**

```
w.override(coupling=..., ignore_failures=...).this()
```

Or, again, shorter:
```
w(coupling=..., ignore_failures=...).this()
```

**Comment by tmontes** [comment]
Additional notes:
- Related to #24.
- "coupling" may indeed not be the best "base word" in this scenario: use "returns" or "results", instead? Needs thought...

**Comment by tmontes** [comment]
API changes and test refactoring ready, from #39.

Next:
* Actually split the concepts and expand the tests.

**Comment by tmontes** [comment]
With the latest PRs and the recently merged PR #44, this is nearly complete:
* Concepts are separated in #41.
* Tests are complete (from #41 and #44).

Pending:
* Document the test generation/synthesis code which may not be immediately obvious.

**Comment by tmontes** [comment]
PR #46, just merged, documents test generation. This can now be closed.
